pennylane/labs/resource_estimation/resource_tracking.py

```diff
@@ -0,0 +1,316 @@
+
+    >>> res = re.get_resources(my_circuit)()
+    >>> print(res)
+    wires: 3
+    gates: 202
+    gate_types:
+    {'Hadamard': 5, 'CNOT': 10, 'T': 187}
+
+    .. details::
+        :title: Usage Details
+
+        Users can provide custom gatesets to track resources with. Consider :code:`my_circuit()` from above:
+
+        >>> my_gateset = {"Hadamard", "RX", "RY", "QFT", "CNOT"}
+        >>> print(re.get_resources(my_circuit, gate_set = my_gateset)())
+        wires: 3
+        gates: 6
+        gate_types:
+        {'Hadamard': 2, 'CNOT': 1, 'RX': 1, 'RY': 1, 'QFT': 1}
+
+        We can also obtain resources for individual operations and quantum tapes in a similar manner:
+
+        >>> op = re.ResourceRX(1.23, 0)
+        >>> print(re.get_resources(op))
+        wires: 1
+        gates: 17
+        gate_types:
+        {'T': 17}
+
+        Finally, we can modify the config values listed in the global :code:`labs.resource_estimation.resource_config`
+        dictionary to have finegrain control of how the resources are computed.
+
+        >>> re.resource_config
+        {'error_rx': 0.01, 'error_ry': 0.01, 'error_rz': 0.01}
+        >>>
+        >>> my_config = copy.copy(re.resource_config)
+        >>> my_config["error_rx"] = 0.001
+        >>>
+        >>> print(re.get_resources(op, config=my_config))
+        wires: 1
+        gates: 21
+        gate_types:
+        {'T': 21}
+
+    """
+
+    raise TypeError(
+        f"Could not obtain resources for obj of type {type(obj)}. obj must be one of Operation, Callable or QuantumScript"
+    )
+
+
+@get_resources.register
+def resources_from_operation(
+    obj: Operation, gate_set: Set = DefaultGateSet, config: Dict = resource_config
+) -> Resources:
+    """Get resources from an operation"""
+
+    if isinstance(obj, ResourceOperator):
+        cp_rep = obj.resource_rep_from_op()
+
+        gate_counts_dict = defaultdict(int)
+        _counts_from_compressed_res_op(cp_rep, gate_counts_dict, gate_set=gate_set, config=config)
+        gate_types = _clean_gate_counts(gate_counts_dict)
+
+        n_gates = sum(gate_types.values())
+        return Resources(num_wires=len(obj.wires), num_gates=n_gates, gate_types=gate_types)
+
+    res = Resources()  # TODO: Add implementation here!
+    return res
+
+
+@get_resources.register
+def resources_from_qfunc(
+    obj: Callable, gate_set: Set = DefaultGateSet, config: Dict = resource_config
+) -> Callable:
+    """Get resources from a quantum function which queues operations!"""
+
+    @wraps(obj)
+    def wrapper(*args, **kwargs):
+        with AnnotatedQueue() as q:
+            obj(*args, **kwargs)
+
+        operations = tuple(op for op in q.queue if not isinstance(op, MeasurementProcess))
+        compressed_res_ops_lst = _operations_to_compressed_reps(operations)
+
+        initial_gate_set = set.union(gate_set, _StandardGateSet)
```

Comment:
Why use `_StandardGateSet` instead of `DefaultGateSet`?

Reply:
See the answer below. Essentially we want to track gates at an intermediate stage first before we go all the way down to the leaf nodes.